### ISSUE TITLE
Bogus behaviour of DirtyStateEditorSupport.UpdateEditorStateJob.unloadAffectedResourcesAndReparseDocument vs OnChangeEvictingCache.CacheAdapter.isSemanticStateChange

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-eclipse#1697]
Given this grammar

```
Model:
	(persons+=Person|greetings+=Greeting)*;
	
Greeting:
	'Hello' to=[Person] '!';
	
Person: "Person" name=ID;	
```

and this Validator

```
public class MyDslValidator extends AbstractMyDslValidator {
	
	@Inject
	private IResourceScopeCache cache;
	
	@Check
	public void checkModel(Model m) {
		List<Person> list = cache.get("somekey", m.eResource(), new Provider<List<Person>>() {

			@Override
			public List<Person> get() {
				List<Person> result = new ArrayList<>();
				for (Greeting g : m.getGreetings()) {
					result.add(g.getTo());
				}
				return result;
			}});
		for (Person p : list) {
			System.out.println(p.eResource()); // should not be null
		}
	}
	
}
```

having these model files

```a.mydsl
Person X
Person Y
Person Z
``` 

```b.mydsl
Hello X!
Hello Y!
Hello Z!
```

doing a clean build will print for the validations in ValidationJob for the editor of the second file

```
null
null
null
```

instead of the expected
```
org.eclipse.xtext.linking.lazy.LazyLinkingResource@39ffc860 uri='platform:/resource/a/a.mydsl'
org.eclipse.xtext.linking.lazy.LazyLinkingResource@39ffc860 uri='platform:/resource/a/a.mydsl'
org.eclipse.xtext.linking.lazy.LazyLinkingResource@39ffc860 uri='platform:/resource/a/a.mydsl'
```

this is happeing cause the a.mydsl will be unloaded but the cache for b.mydsl never will be cleared

**Comment by cdietrich** [comment]
likely caused by this change: https://github.com/eclipse/xtext-core/commit/8990d9c55de3c3a904ac3a0187ad6397fb688ae8#diff-a81b907ab8324e809119b2057dc7203e2400240dd3a2b8745699a62e3666ffe7

**Comment by cdietrich** [comment]
interestingly the change evict seems to happen "after" the check ran with the implicit validation that resolves proxy.
maybe there also was an order change how validations are executed

```
java.lang.Exception
	at org.eclipse.xtext.util.OnChangeEvictingCache$CacheAdapter.isSemanticStateChange(OnChangeEvictingCache.java:320)
	at org.eclipse.xtext.util.OnChangeEvictingCache$CacheAdapter.notifyChanged(OnChangeEvictingCache.java:273)
	at org.eclipse.emf.common.notify.impl.BasicNotifierImpl.eNotify(BasicNotifierImpl.java:424)
	at org.eclipse.emf.ecore.resource.impl.ResourceImpl.load(ResourceImpl.java:1583)
	at org.eclipse.emf.ecore.resource.impl.ResourceImpl.load(ResourceImpl.java:1342)
	at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.demandLoad(ResourceSetImpl.java:259)
	at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.demandLoadHelper(ResourceSetImpl.java:274)
	at org.eclipse.xtext.resource.XtextResourceSet.getResource(XtextResourceSet.java:266)
	at org.eclipse.xtext.resource.SynchronizedXtextResourceSet.getResource(SynchronizedXtextResourceSet.java:25)
	at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.getEObject(ResourceSetImpl.java:220)
	at org.eclipse.emf.ecore.util.EcoreUtil.resolve(EcoreUtil.java:209)
	at org.eclipse.emf.ecore.util.EcoreUtil.resolve(EcoreUtil.java:269)
	at org.eclipse.emf.ecore.impl.BasicEObjectImpl.eResolveProxy(BasicEObjectImpl.java:1516)
	at org.xtext.example.mydsl9.myDsl.impl.GreetingImpl.getTo(GreetingImpl.java:75)
	at org.xtext.example.mydsl9.myDsl.impl.GreetingImpl.eGet(GreetingImpl.java:120)
	at org.eclipse.emf.ecore.impl.BasicEObjectImpl.eGet(BasicEObjectImpl.java:1050)
	at org.eclipse.emf.ecore.impl.BasicEObjectImpl.eGet(BasicEObjectImpl.java:1042)
	at org.eclipse.emf.ecore.util.EContentsEList$FeatureIteratorImpl.hasNext(EContentsEList.java:439)
	at org.eclipse.emf.ecore.util.EObjectValidator.validate_EveryReferenceIsContained(EObjectValidator.java:782)
	at org.eclipse.emf.ecore.util.EObjectValidator.validate_EveryDefaultConstraint(EObjectValidator.java:355)
	at org.eclipse.emf.ecore.util.EObjectValidator$DynamicEClassValidator.validate(EObjectValidator.java:1426)
	at org.eclipse.emf.ecore.util.EObjectValidator.validate(EObjectValidator.java:333)
	at org.eclipse.xtext.validation.CompositeEValidator.validate(CompositeEValidator.java:151)
	at org.eclipse.emf.ecore.util.Diagnostician.doValidate(Diagnostician.java:299)
	at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:245)
	at org.eclipse.xtext.validation.CancelableDiagnostician.validate(CancelableDiagnostician.java:41)
	at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:200)
	at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:142)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:147)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:125)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:91)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:92)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob$1.exec(ValidationJob.java:1)
	at org.eclipse.xtext.util.concurrent.CancelableUnitOfWork.exec(CancelableUnitOfWork.java:27)
	at org.eclipse.xtext.util.concurrent.WrappingCancelableUnitOfWork.exec(WrappingCancelableUnitOfWork.java:58)
	at org.eclipse.xtext.util.concurrent.CancelableUnitOfWork.exec(CancelableUnitOfWork.java:27)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:70)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:525)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:497)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:136)
	at org.eclipse.xtext.util.concurrent.IReadAccess.tryReadOnly(IReadAccess.java:50)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.createIssues(ValidationJob.java:87)
	at org.eclipse.xtext.ui.editor.validation.ValidationJob.run(ValidationJob.java:68)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

**Comment by cdietrich** [comment]
@szarnekow  any idea why we dont relink/unload the main resource in the editor but only affected others in the resourceset?

**Comment by szarnekow** [comment]
Hmm the primary resource should be processed via `org.eclipse.xtext.ui.editor.reconciler.XtextDocumentReconcileStrategy.doReconcile(IRegion)` - maybe forceReconcile does not enough, e.g. a dummy change is scheduled. Maybe the dummy change is not enough to clear the wrong cache entries?

**Comment by cdietrich** [comment]
yes it looks like. when i revert the change like
```

import org.eclipse.emf.common.notify.Notification;
import org.eclipse.emf.ecore.resource.Resource.Diagnostic;
import org.eclipse.xtext.util.OnChangeEvictingCache;

import com.google.inject.Singleton;

@Singleton
public class PatchedOnChangeEvictingCache extends OnChangeEvictingCache {
	
	@Override
	protected CacheAdapter createCacheAdapter() {
		return new CacheAdapter() {
			@Override
			protected boolean isSemanticStateChange(Notification notification) {
				return !notification.isTouch() && !(notification.getNewValue() instanceof Diagnostic) && !(notification.getOldValue() instanceof Diagnostic);
			}
		};
	}

}
```
it does work again

**Comment by cdietrich** [comment]
doing a replace dummy change does not always seems to help. there still seems to be a dependence on validator order

```
public void forceReconcile() {
		if(editor != null && editor.getDocument() != null) {
			IXtextDocument document = editor.getDocument();
			final DocumentEvent dummyEvent;
			if (document.getLength()>0) {
				try {
					dummyEvent = new DocumentEvent(document, 0, 1, document.get(0, 1));
				} catch (BadLocationException e) {
					throw new RuntimeException(e);
				}
			} else {
				dummyEvent = new DocumentEvent(document, 0, 0, "");
			}
			handleDocumentChanged(dummyEvent);
		}
	}
``´
or on how the reconciler races against the build (when the stuff is actually unresolvable or not)

**Comment by cdietrich** [comment]
i wonder weather org.eclipse.xtext.resource.XtextResource.updateInternalState(IParseResult) should also clear caches besides errors and co

**Comment by cdietrich** [comment]
e.g. as a patch

```
public class MyXtextResource extends LazyLinkingResource {
	
	@Override
	protected void updateInternalState(IParseResult newParseResult) {
		setParseResult(newParseResult);
		EObject newRootASTElement = getParseResult().getRootASTElement();
		if (newRootASTElement != null && !containsRootElement(newRootASTElement))
			getContents().add(0, newRootASTElement);
		reattachModificationTracker(newRootASTElement);
		IResourceScopeCache cache = getCache();
		if (cache instanceof OnChangeEvictingCache) {
			((OnChangeEvictingCache) cache).clear(this);
		}
		clearErrorsAndWarnings();
		addSyntaxErrors();
		doLinking();
	}
	
	boolean containsRootElement(EObject newRootASTElement) {
		return getContents().contains(newRootASTElement);
	}

}
```

question would be: where in the chain is the correct place

**Comment by szarnekow** [comment]
Since the cache clearing used to happen automagically along with clearErrorAndWarnings, it should probably be put there. It should be possible to override the clear routing for the cache though, e.g with a new protected method `clearCache()` on the resource.

**Comment by cdietrich** [comment]
@szarnekow do you know why the clear(resource) in the OnChangeEvicitingCache does the getOrCreate?

```
	protected void clearCache() {
		IResourceScopeCache cache = getCache();
		if (cache instanceof OnChangeEvictingCache) {
			CacheAdapter adapter = (CacheAdapter) EcoreUtil.getAdapter(eAdapters(), CacheAdapter.class);
			if (adapter != null) {
				adapter.clearValues();
			}
		}
	}
``` 
would not create a cache if not needed, 

```
	protected void clearCache() {
		IResourceScopeCache cache = getCache();
		if (cache instanceof OnChangeEvictingCache) {
			cache.clear(this);
		}
	}
```
would be compatible, but might create unnecessary adapters

**Comment by cdietrich** [comment]
Fixed in 2.26
